pkgs/build-support/fetchgit/nix-prefetch-git

```diff
@@ -96,7 +98,7 @@ for arg; do
         case $argfun in
             set_*)
                 var=${argfun#set_}
-                eval $var=$arg
+                eval $var="\"$arg\""
```

Comment:
This doesn't seem necessary. If you want to make it "correct" you could use `printf %q`, but the real solution is to replace this code by something simple, like a loop that pops arguments using `shift`.
I'll ignore it for now.

Reply:
This is necessary because --sparse-checkout arg may contain "\n", without this change, the following test fails.

```
~/src/nixpkgs master* ⇣⇡
❯ git diff
diff --git a/pkgs/build-support/fetchgit/nix-prefetch-git b/pkgs/build-support/fetchgit/nix-prefetch-git
index f940d10b2ae..be595e59b0c 100755
--- a/pkgs/build-support/fetchgit/nix-prefetch-git
+++ b/pkgs/build-support/fetchgit/nix-prefetch-git
@@ -98,7 +98,7 @@ for arg; do
         case $argfun in
             set_*)
                 var=${argfun#set_}
-                eval $var="\"$arg\""
+                eval $var=$arg
                 ;;
         esac
         argfun=""

~/src/nixpkgs master* ⇣⇡
❯ nix repl .
Welcome to Nix version 2.5pre20211007_844dd90. Type :? for help.

Loading '.'...
Added 15361 variables.

nix-repl> :b fetchgit.tests.sparseCheckout                    
error: builder for '/nix/store/cnh4q2g2vmxl6l4cjq28677hqdxrwxbx-nix-source-salted-bydqcp0cva5n.drv' failed with exit code 127;
       last 2 log lines:
       > exporting https://github.com/NixOS/nix (rev 9d9dbe6ed05854e03811c361a3380e09183f4f4a) into /nix/store/yw6ycpfg088q7jswqjgjj19724drl3ji-nix-source-salted-bydqcp0cva5n
       > /nix/store/k4wdi6yqznx8qww12v715sz46vd4p1jf-nix-prefetch-git: line 101: tests: command not found
       For full logs, run 'nix log /nix/store/cnh4q2g2vmxl6l4cjq28677hqdxrwxbx-nix-source-salted-bydqcp0cva5n.drv'.
```

The `tests` comes from:

```
    sparseCheckout = ''
      src
      tests
    '';
```

`eval` without quote leads to code injection.

Reply:
How about something like `eval "$var=$(printf %q "$arg")"`

It could also append to the previous value, so it can be used multiple times on the command line. I guess that'd be `append_*` although I think the argument parsing should be turned into a simple loop that consumes one or more arguments per iteration. Much simpler than this weird state machine business.

Reply:
Done

Reply:
`declare "$var=$arg"` ...